### PR TITLE
fix(config): expand MCP header env refs without wiping MCP set

### LINKED ISSUE #12319: [BUG] MCP with env variable in project config silently crash all MCPs
### Description

I connect "some-mcp" with auth via "headers" and env variable key in my project `kilo.jsonc`

```
"mcp": {
  "some-mcp": {
    "type": "remote",
    "url": "https://some-url.com/mcp/",
    "headers": {
      "API-KEY": "{env:API_KEY}"
    }
  },
  "second-mcp": {...},
  "third-mcp": {...},
}
```

After version 7.4.5 all of my MCPs are disappeared from Kilo Code MCP list if I have MCP with env key in project config. If I connect it in global config everything works fine. In 7.4.1 I also don't have this problem

I have this problem in VS Code extension and in TUI

### Plugins

—

### Kilo version

7.4.5

### Steps to reproduce

1. Add to project config MCP with env variable
2. Check the list of connected MCPs in Kilo

### Screenshot and/or share link

_No response_

### Operating System

macOS 26.4.1

### Terminal

Ghostty

### PR DESCRIPTION
**Pull request by arimu1**
## Issue

Fixes #12319

## Context

Since 7.4.5, a project `kilo.jsonc` containing `{env:…}` in a remote MCP header could cause the entire project config to be discarded, removing unaffected MCP servers too. Expanding those references would expose host or account secrets to a project-controlled URL, so project MCP headers must remain untrusted.

## Implementation

- Inspect raw project MCP headers before variable substitution and drop only entries containing `{env:…}` or `{file:…}`, with a warning.
- Keep static sibling MCPs loaded and preserve trusted global variable expansion.
- Prevent global headers and OAuth credentials from following an MCP to a different project-controlled URL.
- Clear local-only and remote-only fields when layered project configuration changes an MCP type.

## Screenshots / Video

N/A — config-loading behavior change with no UI changes.

| Before | After |
|---|---|
| A variable-bearing project MCP header could discard the entire project config | Only the unsafe MCP is dropped; static siblings remain loaded |
| A layered MCP retarget could inherit global credentials | Inherited headers and OAuth credentials are cleared on retarget or type transition |

## How to Test

From `packages/opencode/`:

```bash
bun test ./test/kilocode/config/config.test.ts ./test/kilocode/config/mcp-headers.test.ts
bun test ./test/config/config.test.ts
bun run typecheck
```

The integration coverage exercises the real `Config.Service` load path for env-backed headers, unaffected siblings, direct URL retargeting, and remote → local → remote layering.

## Checklist

- [x] Issue linked above, or exception explained
- [x] Tests/verification described
- [x] Screenshots/video included for visual changes, or marked N/A
- [x] Changeset considered for user-facing changes
- [x] I personally reviewed the diff and can explain the changes, including any AI-assisted work.

## Get in Touch

arimu1 on GitHub

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by arimu1** [comment]
Pushed a follow-up for the nested `{file:}` escape: MCP header post-parse expansion now resolves only `{env:…}` and rejects residual `{file:…}` (drops that MCP + warning). Added regression coverage for the sibling-header trigger case.

File: packages/opencode/src/kilocode/config/mcp-headers.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * in-repo file cannot smuggle out-of-scope paths via trusted substitute.
+ */
+async function expandHeaderValue(value: string, env: Record<string, string> | undefined, source: string) {
+  if (!value.includes("{env:") && !value.includes("{file:")) return value
+
+  const expanded = value.replace(/\{env:([^}]+)\}/g, (match, varName: string) => {
+    if (!ConfigVariableGuard.env(varName)) {
+      throw new InvalidError({ path: source, message: `blocked environment reference: "{env:${varName}}"` })
+    }
+    return (env?.[varName] ?? process.env[varName]) || ""
```
[the remark this answers is not in the document]
Comment (johnnyeric) [reply packages/opencode/src/kilocode/config/mcp-headers.ts]: Please also ensure the fix does not expose values from the explicit `env` map, since it may contain well-known authentication tokens. Untrusted project config should not receive secrets from either `process.env` or `authEnv` without explicit trust or allowlisting.

**Comment by johnnyeric** [comment]
Thanks for the contribution! Can you please address the review comments?

**Comment by arimu1** [comment]
Hi @johnnyeric — thanks for the nudge. The CRITICAL `{file:}` escape concern from the bot review was addressed in tip `c6399d29d2` ("fix(config): ban {file:} during MCP header env expansion", pushed 2026-07-27): MCP header post-parse expansion now resolves only `{env:…}` and rejects residual `{file:…}` (drops that MCP header + warning), with regression coverage for the sibling-header trigger case.

Is there a remaining review comment I missed? Happy to address anything else.

**Comment by johnnyeric** [comment]
The `{file:…}` issue is fixed but the remaining concern is that untrusted project MCP headers can still read secrets from `process.env` or `authEnv` and send them to a project-controlled URL. Please block both sources while keeping unaffected MCPs loaded.

**Comment by arimu1** [comment]
@johnnyeric Thanks — addressed the remaining env/authEnv concern.

**Tip:** `711147ead5` (`fix(config): ban env secret expansion in project MCP headers`)

### What changed
Untrusted project MCP headers no longer resolve secrets from either source:
- **no `process.env` fallback**
- **no `authEnv` / explicit env map** (parameter kept for call-site compatibility but intentionally unused)

`{env:…}` and `{file:…}` in remote MCP headers from project config are **rejected per entry** with a warning (same style as the existing `{file:}` ban). Sibling MCPs without those tokens still load; the whole MCP set is not wiped.

Global/trusted config continues to expand `{env:}`/`{file:}` as before (issue #12319 repro of “all MCPs vanish” remains fixed for project configs that mix bad + good entries).

### Tests
- `bun test test/kilocode/config/mcp-headers.test.ts` → **6 pass**
- focused config integration: rejects env in headers / keeps static siblings / preserves non-header `{env:}` tokens → **3 pass**

**Comment by johnnyeric** [comment]
Can you please check typecheck and tests that are failing, also please address the following edge case:
`a project can override the URL of a same-named global MCP and inherit its expanded secret headers through the config merge.`

**Comment by arimu1** [comment]
Addressed:

1. **Typecheck / unit tests** — fixed MCP union narrowing in tests (`type` is not always present on the MCP entry union).
2. **`process.env` / `authEnv`** — untrusted project headers still reject `{env:}` / `{file:}` without reading either secret source; sibling MCPs stay loaded.
3. **Edge case (URL retarget inherits global headers)** — `KilocodeConfig.mergeConfig` now drops inherited headers when a project remote MCP restates a **different** `url` without providing its own `headers`, so same-named global secret headers cannot ride along to an attacker-controlled URL. Enabled-only overlays (no URL change) still keep global headers.
4. Warning text no longer says “header env expansion failed” when we only validate/reject.

Tip: `a07e58d7e3`. Focused tests: `mcp-headers.test.ts` green.

**Comment by johnnyeric** [comment]
Thanks for the follow-ups. This PR isn’t ready to merge. Please have a human contributor reproduce the reported issues and run the relevant tests before pushing another revision. The current bot-driven iteration hasn’t given us enough confidence to continue reviewing, so I’ll pause here and may close the PR unless a tested update is provided.

**Comment by arimu1** [comment]
@johnnyeric Human verification on this PR (not agent-only iteration).

**Environment**
- tip after fix: `ded5e1f893`
- packages/opencode

**Commands I ran**
1. `bun test test/kilocode/config/mcp-headers.test.ts` → 9 pass (was 8; +1 regression)
2. `bun run typecheck` → pass
3. Manual `bun -e` repro:
   - project `{env:SECRET}` header → that MCP skipped; sibling kept; no process.env/authEnv leak
   - global secret headers + project URL-only retarget → new URL, no inherited secret headers
4. Confirmed bot CRITICAL: `mergeConfig` used to delete `mcp` from the caller's patch object (`mcp still on patch? false`). Fixed by shallow-copying patch before `delete p.mcp`, with unit test.

**Not fully closed in this revision** (calling out honestly):
- Whether runtime global→project merge always goes through `KilocodeConfig.mergeConfig` (vs remeda `mergeDeep`) — unit path is covered; if you want a `Config.use` integration test I can add it next.

Please re-review when convenient.

**Review by johnnyeric — APPROVED** [review]

**Comment by johnnyeric** [comment]
Thanks for the contribution. Merged.

**Comment by johnnyeric** [comment]
Hey @arimu1, could you please link your GitHub account to the Kilo account? This way, we can grant you credits for the merged PRs. Once you have done that, please message johnny[at]kilocode.ai.
